docs/how_tos/theming.md

```diff
@@ -0,0 +1,40 @@
+# Theming support with Paragon
+
+Status: Draft
+
+This document serves as a guide to using `@edx/frontend-platform` to support MFE theming with Paragon using theme CSS loaded externally (e.g., from a CDN). By serving CSS from externally, consuming applications of Paragon no longer need to be responsible for compiling the theme SCSS to CSS themselves and instead use a pre-compiled CSS file instead. In doing so, this allows making changes to the Paragon theme without needing to necessarily re-build and re-deploy all consuming applications. We would also get a meaningful gain in performance as loading the compiled theme CSS from an external CDN means micro-frontends (MFEs) can include cached styles instead of needing to load essentially duplicate theme styles as users navigate across different MFEs.
+
+## Theme URL configuration
+
+Paragon supports 2 mechanisms for configuring the Paragon theme URLs:
+* Environment variable configuration
+* Runtime configuration
+
+The Paragon theming extension to dynamically load external theme CSS prefers the theme configuration in the runtime config over the environment variable configuration.
+
+### Environment variable configuration
+
+The standard way to configure MFEs is to use environment variables specific to the application environment they are running in. For example, during local development, environment variables are defined and loaded via the `.env.development` file.
+
+Two new environment variables are exposed to configure the Paragon theme URLs:
+* `PARAGON_THEME_CORE_URL`. This URL represents the foundational theme styles provided by Paragon's `core.css` file.
+* `PARAGON_THEME_VARIANTS_LIGHT_URL`. This URL represents the light theme variant specific styles provided by Paragon's `light.css` file.
+
+### Runtime configuration
+
+`@edx/frontend-platform` additionally supports loading application configuration from an API at runtime rather than environment variables. For example, in `edx-platform`, there is an API endpoint for MFE runtime configuration at `http://localhost:18000/api/mfe_config/v1`. The application configuration may be setup via Django settings as follows:
+
+```
+ENABLE_MFE_CONFIG_API = True
+MFE_CONFIG = {}
+MFE_CONFIG_OVERRIDES = {
+    "profile": {
+        "PARAGON_THEME_URLS": {
+            'core': 'https://cdn.jsdelivr.net/npm/@edx/paragon@21.0.0-alpha.15/dist/paragon.css',
```

Comment:
Would it make sense to support variable substitutions here? i.e. the URL could be:

``https://cdn.jsdelivr.net/npm/@edx/paragon@$paragonVersion/dist/paragon.css``

This way with each upgrade you don't need to update this value here in a completely different place.

Reply:
The last version of Paragon Alpha creates a folder instead of a file (core, and theme variant)

- @edx/paragon
    - styles
       - css
          - core
              - custom-media-breakpoints.css
              - index.css
              - variables.css
         - themes
           - light
               -  index.css
                - utility-classes.css
                - variables.css
               
I was wondering if this implementation should take that into consideration or should be resolved by the jsdelivr capacity to load more than one file per request https://www.jsdelivr.com/documentation#id-combine-multiple-files?

Reply:
> Would it make sense to support variable substitutions here? i.e. the URL could be: https://cdn.jsdelivr.net/npm/@edx/paragon@$paragonVersion/dist/paragon.css

@xitij2000 Can you expand a bit more on how you think the variable substitution would  work here? Where would `$paragonVersion` come from?

Reply:
> > Would it make sense to support variable substitutions here? i.e. the URL could be: https://cdn.jsdelivr.net/npm/@edx/paragon@$paragonVersion/dist/paragon.css
> 
> @xitij2000 Can you expand a bit more on how you think the variable substitution would work here? Where would `$paragonVersion` come from?

The idea with variable substitution is that if you upgrade the version of paragon in the MFE it should automatically inject its version of paragon into this URL instead of requiring you to update the MFE configuration is site-configuration, or in settings. 

The version of Paragon to be filled in here is known at build time since it's in package.json so it can just be a static variable auto-discovered from the package.json file.

Reply:
@xitij2000 That's a great suggestion, which should definitely mitigate the concern of needing to keep the CDN theme url versions in sync with the locally installed Paragon version.

To be able to support this, I believe we'd need to expose a `PARAGON_VERSION` environment variable from `@edx/frontend-build`, e.g.: https://github.com/openedx/frontend-build/pull/365

Reply:
[inform] I pushed up a commit using the above linked version of `@edx/frontend-build`.

Reply:
Assuming I am not incorrect, this would lead to style shift, right? because the flow as I understnad would as 
- The HTML of the is MFE is served 
- Browser start fetching resoucrs app.js .app.css ...etc
- Now concurrently browser would 
  - Execuse the app.js _of which it fetch the dynamic config api_ 
  -  Start rendering the screen
- Now if dynamic config from previous step would chane the style, there would be stlyle shift i.e. _user would see a theme style and then **suddenly** the theme/style would change_. 

Am I right, with the timeline of events above?, either cases I think the use PARAGON_VERSION is valuable, but we probably shall ask if sudden style change is okay and if not how to resolve it...

Reply:
The problem I am referring to is exactly this https://css-tricks.com/content-jumping-avoid/

Reply:
Reading into the thread, I think my comment above, is related even if we don't define `PARAGON_VERSION`. i.e even if just define `PARAGON_THEME_URLS`, _wouldn't the MFE at first load the deafult theme that was used/set when the MFE is built, before it get a chance to update it using via the dynamic config API_?.

Reply:
@ghassanmas I believe the intent here is that the MFE would no longer import any of Paragon's theme SCSS, in favor of having `@edx/frontend-platform` inject it via `AppProvider` at runtime, or otherwise fallback to the CSS included in the locally installed Paragon verison.

In this case, the MFE isn't shown until the externally loaded CSS files are injected/loaded by the browser, similar to how the MFE isn't rendered until the request to fetch the dynamic MFE config API is resolved.

See the below graphic for the flow in this proposed implementation for injecting/loading the Paragon theme URLs (I'll plan on adding a similar graphic to the ADR in this PR as well):

![image](https://github.com/openedx/frontend-platform/assets/2828721/573ba78d-2b37-47e0-82a9-4ad9b89de47d)